src/vs/editor/contrib/format/format.ts

```diff
@@ -169,27 +169,62 @@ export async function formatDocumentRangesWithProvider(
 		}
 	}
 
-	const allEdits: TextEdit[] = [];
+	const computeEdits = async (range: Range) => {
+		const rawEdits = await provider.provideDocumentRangeFormattingEdits(
+			model,
+			range,
+			model.getFormattingOptions(),
+			cts.token
+		);
+		return (await workerService.computeMoreMinimalEdits(model.uri, rawEdits)) || [];
+	};
+
+	const hasIntersectingEdit = (a: TextEdit[], b: TextEdit[]) => {
+		for (let edit of a) {
+			for (let otherEdit of b) {
+				if (Range.intersectRanges(edit.range, otherEdit.range)) {
+					return true;
+				}
+			}
+		}
+		return false;
+	};
+
+	const rawEditsList: TextEdit[][] = [];
 	try {
 		for (let range of ranges) {
-			const rawEdits = await provider.provideDocumentRangeFormattingEdits(
-				model,
-				range,
-				model.getFormattingOptions(),
-				cts.token
-			);
-			const minEdits = await workerService.computeMoreMinimalEdits(model.uri, rawEdits);
-			if (minEdits) {
-				allEdits.push(...minEdits);
-			}
+			rawEditsList.push(await computeEdits(range));
 			if (cts.token.isCancellationRequested) {
 				return true;
 			}
 		}
+
+		for (let i = 0; i < ranges.length; ++i) {
+			for (let j = i + 1; j < ranges.length; ++j) {
+				if (cts.token.isCancellationRequested) {
+					return true;
+				}
+				if (hasIntersectingEdit(rawEditsList[i], rawEditsList[j])) {
+					// Merge ranges i and j into a single range, recompute the associated edits
```

Comment:
If there are 1000 modified ranges and the formatter reports 1000 edits for each format range request (which at least prettier does sometimes even for simple format range requests), you run 1000^4 operations here.

I think it should be fine if you just test if any two ranges touch the same line. Then you could use a set of touched line numbers and get an algorithm than runs in O(line numbers).

Reply:
You could also compute more minimal edits when you are done. There should be a guarantee that computing more minimal edits doesn't out grow the range (and therefore doesn't create overlap). Usually, extensions compute very few edits and making them more minimal often creates many, many edits

Reply:
I now compute minimal edits only at the end.
I also added a quick exit to the `hasIntersectingEdit` function.

> I think it should be fine if you just test if any two ranges touch the same line. Then you could use a set of touched line numbers and get an algorithm than runs in O(line numbers).

It's not that easy, not only do I need to know that 2 edits intersect, but I also need to know which 2 ranges are at the original of these, and the list of all edits produced by these 2 ranges so I can merged them. 
So I need to iterate all combination of ranges in any case. That's already O(n^2). Then I need to know if the edits intersect. That's O(m^2). I guess the overall complexity of O(n^2*m^2) is what you refer to as O(n^4).
I don't think I can do much better while retaining correctness and completeness, at least not at the cost of an overly complicated algorithm (usefulness vs maintainability). 

Now that minimal edits are only computed at the end, I expect the problem size to stay rather small, which means complexity should not be an issue that much.
If you're worried I still have a few tricks:
1. Quick exit on `hasIntersectingEdits`, brings down O(m^2) down to O(m) in most cases (not really O then, agreed), worst case is not affected, but we improve most use cases while retaining correctness. (already implemented)
2. Assume only consecutive ranges can produce intersecting edits: although not guaranteed, I don't see any case where a sane formatter would not respect this. That would bring O(n^2) down to O(n) at the cost of potentially not merging 2 ranges...which would be just as good as what we're doing before this PR
3. Skip this algorithm if the problem size is too big. Format range merging would still work as long as there are not too many edits. Again, strictly not worse than the current state of things. The problem with this is finding the correct threshold.

Let me know what you think, or if you have any other ideas

Reply:
>  I guess the overall complexity of O(n^2*m^2) is what you refer to as O(n^4).

Yes, if the number of edits equals the number of modified lines, we have n = m. Can you try out your algorithm on [checker.ts](https://github.com/microsoft/TypeScript/blob/main/src/compiler/checker.ts) when you edit a lot of lines? Can you compare the time with how long it takes without this change (I guess just formatting a single range in that file with prettier might be already slow)?

> Quick exit on hasIntersectingEdits, brings down O(m^2) down to O(m) in most cases

That is a good idea!

> Skip this algorithm if the problem size is too big. Format range merging would still work as long as there are not too many edits. 

I think we should do this if it becomes slow for really large files.

Reply:
An easy trick would be to count of often we have spliced and merged ranges and simply stop in case a threshold has been reached. Tho, I would suggest we wait for that bug to be reported. I am happy with the changes and we can still improve it later.